core.lua

```diff
@@ -536,6 +536,123 @@ do
 			return count
 		end
 	end
+
+	-- ================
+	-- = Leaderboards =
+	-- ================
+	local function LearderboardRealmData (region, realmName, dungeonId)
+		local realmData = ns.leaderboardData[region][realmName]
+
+		-- Connected realms store the name of another realm
+		-- where the data is stored
+		if type(realmData) == "string" then
+			realmData = ns.leaderboardData[region][realmData]
+		end
+
+		return realmData[dungeonId]
+	end
+
+	local function MinimumKeyLevel (dungeon, realmName)
+		realmName = (realmName or GetNormalizedRealmName()):gsub("%s+", "")
+
+		local lowestData = LearderboardRealmData(ns.PLAYER_REGION, realmName, dungeon["id"])
+		return lowestData["mythicLevel"] or 2, lowestData["rank"] or 0, lowestData["upgrades"] or 0
+	end
+
+	-- Returns a string indicating keystone level and upgrades,
+	-- with the level number in the provided colour.
+	local function FormatMinimumKeystone (minimumKeyLevel, upgrades, color)
+		local keyColor = color or ""
+		local stars = GetStarsForUpgrades(upgrades, true)
+
+		return format("%s%s%s", stars, keyColor, minimumKeyLevel)
+	end
+
+	function AddMinimumKeyLevelTooltipForParty(tooltip, dungeon, contextLevel, additionalPlayer, additionalPlayerTitle)
+		local COLOR_GREEN = "|cff00ff00"
+		local COLOR_RED = "|cffff0000"
+		local COLOR_ORANGE = "|cffFF4500"
+
+		local rows = {}
+
+		-- Collect players
+		local n = GetNumGroupMembers()
+		for i = 0, n do
+			local unit = i == 0 and "player" or "party" .. i
+			local name, realm = GetNameAndRealm(unit)
+			table.insert(rows, {title = name, realm = realm})
+		end
+
+		if additionalPlayer then
+			local name, additionalRealm = GetNameAndRealm(additionalPlayer)
+			if additionalRealm then
+				local title = additionalPlayerTitle or name
+				table.insert(rows, {title = title, realm = additionalRealm, additional = true})
+			end
+		end
+
+		-- Add leaderboard data
+		for _, row in ipairs(rows) do
+			local keyLevel, rank, upgrades = MinimumKeyLevel(dungeon, row.realm)
+			row.level = keyLevel
+			row.upgrades = upgrades
+			row.rank = rank
+		end
+
+		table.sort(rows, function(a, b) return a.level < b.level end)
+
+		-- Find lowest rank from collected realms
+		local minRow = nil
+
+		for _, row in ipairs(rows) do
+			if not minRow or row.rank < minRow.rank or row.level < minRow.level then
+				minRow = row
+			end
+		end
+
+		local LEADERBOARD_CAPACITY = 500
+		local LEADERBOARD_NEARING_CAPACITY = LEADERBOARD_CAPACITY * 0.75
+
+		-- We can’t meaningfully display a minimum if the leaderboard is not yet full,
+		-- since the minimum will always be 2.
+		local show_level = false
+		local status
+
+		if minRow.rank == LEADERBOARD_CAPACITY then
+			status = COLOR_RED.."Full"
+			show_level = true
+		elseif minRow.rank >= LEADERBOARD_NEARING_CAPACITY then
+			status = COLOR_ORANGE.."Almost Full"
+		else
+			status = COLOR_GREEN.."Open"
+		end
+
+		tooltip:AddDoubleLine("M+ Dungeon Leaderboards", status)
```

Comment:
Maybe only display this line if either `show_level` or `contextLevel` is true (to avoid having only the title and nothing below) (not even sure that's even possible)